kernel/work.c

```diff
@@ -0,0 +1,966 @@
+	} else {
+		/* Caught the item before being invoked; complete the
+		 * cancellation now.
+		 */
+		finalize_cancel_locked(work);
+	}
+
+	k_spin_unlock(&lock, key);
+
+	return ret;
+}
+
+/* Work handler has been called and is about to go idle.
+ *
+ * If the work is being canceled this will notify anything waiting
+ * for the cancellation.
+ *
+ * Invoked by work queue thread.
+ * Takes and releases lock.
+ * Reschedules via finalize_cancel_locked
+ *
+ * @param work work that is in running state
+ */
+static inline void work_clear_running(struct k_work *work)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	/* Clear running */
+	atomic_clear_bit(&work->flags, K_WORK_RUNNING_BIT);
+
+	if (atomic_test_bit(&work->flags, K_WORK_CANCELING_BIT)) {
+		finalize_cancel_locked(work);
+	}
+
+	k_spin_unlock(&lock, key);
+}
+
+/* Loop executed by a work queue thread.
+ *
+ * @param workq_ptr pointer to the work queue structure
+ */
+static void work_queue_main(void *workq_ptr, void *p2, void *p3)
+{
+	struct k_work_q *queue = (struct k_work_q *)workq_ptr;
+
+	while (true) {
+		sys_snode_t *node;
+		struct k_work *work = NULL;
+		k_spinlock_key_t key = k_spin_lock(&queue->lock);
+
+		/* Clear the record of processing any previous work, and check
+		 * for new work.
+		 */
+		node = sys_slist_get(&queue->pending);
+		if (node != NULL) {
+			/* Mark that there's some work active that's
+			 * not on the pending list.
+			 */
+			atomic_set_bit(&queue->flags, K_WORK_QUEUE_BUSY_BIT);
+			work = CONTAINER_OF(node, struct k_work, node);
+		} else if (atomic_test_and_clear_bit(&queue->flags,
+						     K_WORK_QUEUE_DRAIN_BIT)) {
+			/* Not busy and draining: move threads waiting for
+			 * drain to ready state.  The held spinlock inhibits
+			 * reschedule; released threads get their chance when
+			 * this loops back to wait on notify, or the thread
+			 * exits.
+			 *
+			 * We don't touch K_WORK_QUEUE_PLUGGABLE, so getting
+			 * here doesn't mean that the queue will allow new
+			 * submissions.
+			 */
+			(void)z_sched_wake_all(&queue->drainq, 1, NULL);
+		}
+
+		if (work == NULL) {
+			/* Nothing's had a chance to add work since we took
+			 * the queue lock, and We didn't find work nor got
+			 * asked to stop.  Just go to sleep: when something
+			 * happens the work thread will be woken and we can
+			 * check again.
+			 */
+
+			(void)z_sched_wait(&queue->lock, key, &queue->notifyq,
+					   K_FOREVER, NULL);
+			continue;
+		}
+
+		k_spin_unlock(&queue->lock, key);
+
+		if (work != NULL) {
+			k_work_handler_t handler = work->handler;
+
+			__ASSERT_NO_MSG(handler != 0);
+
+			if (work_set_running(work, queue)) {
+				handler(work);
+				work_clear_running(work);
+			}
+
+			/* No longer referencing the work, so we can clear the
+			 * BUSY flag while we yield to prevent starving other
+			 * threads.
+			 */
+			key = k_spin_lock(&queue->lock);
+			atomic_clear_bit(&queue->flags, K_WORK_QUEUE_BUSY_BIT);
+			k_spin_unlock(&queue->lock, key);
```

Comment:
Let's discuss this, because this is a subtlety that I'm not entirely sure I have correct, and have wanted to be able to discuss.

Here's the general concept:

Back in the day of `volatile` the compiler was blocked from optimizing away reads in busy-wait loops like:

    volatile int ready = 0;

    ...
    /* Something will set ready to non-zero */
    while (!ready) { /* spin */}

This worked because `volatile` was attached to the data object.

In [GCC atomics](https://gcc.gnu.org/wiki/Atomic/GCCMM/AtomicSync) the "atomicity" is not attached to the object, but to the operation.  Thus if you do:

    atomic_t ready; /* int ready */
    ...
    /* Something will __atomic_set() or __atomic_cas() ready to non-zero */
    while (!ready) { /* spin */

there is nothing that stops the compiler from optimizing this into:

    if (!ready) { while (true); } /* stay forever */

But the compiler cannot optimize away the repeated checks in:

    while (__atomic_load(&ready, __ATOMIC_SEQ_CST) == 0) { /* spin */

Consequently everywhere I'm manipulating flags that might be accessed outside of a lock I've consistently used the atomic API.  (The sole use outside the lock is in instantaneous state queries like `k_work_pending()`, but I thought avoiding the lock was worthwhile.  I'm less convinced of that now, particularly as you've taught me that the atomic API is not nearly-free.)

In SMP we know `spin_lock()` guarantees atomic sequential consistency because there's an atomic operation involved in the spin lock.  Without SMP all we have is `arch_irq_lock()/unlock()`.

So *if* `spin_lock()` is guaranteed to produce a memory barrier for *both* SMP and non-SMP systems (which in the current implementation means `arch_irq_lock/unlock` are memory barriers), then we can drop all the atomic bit manipulation at the cost only of having to take the lock in order to read the instantaneous state.

I'm quite happy to do that, if that guarantee is there.  It *probably* is because if it wasn't I think all kinds of things could break, but it isn't explicitly stated anywhere I can see.

Does that make sense?

Reply:
Yes.  Both spinlocks and the underlying arch_irq_lock implementations enforce compiler memory barriers always.  It's a fundamental thing, they have to do that or else they don't actually work.

Reply:
Wasn't this what @andrewboie  said 2 weeks ago? https://github.com/zephyrproject-rtos/zephyr/pull/29618#discussion_r556177681 and here https://github.com/zephyrproject-rtos/zephyr/pull/29618#discussion_r556709965

Reply:
For those following along: #32133 suggests I was right to be concerned about this, and that maybe it's good #28902 never got reviewed because the consistency promises it documents aren't actually true.